### ISSUE TITLE
template/preset as i/o database (Old name:how make node trees reusable in easy way)template/preset as just an i/o

### ISSUE THREAD

**Issue by JoseConseco** [JacquesLucke/animation_nodes#412]
Let's assume user have three node trees 1,2,3. 
- nodeTree number 1, does it's own thing and has subprogram **a)** definition, 
- nodeTree nr 2, does it's own thing and  has subprogram **b)** definition
- nodeTree nr 3, does it's own thing and uses subprogram **a)** and*\* b)*\* from above.
  Now when user want append nodeTree 3 into new blend, he is forced to also append tree1 and tree2, just to get subprograms a and b definitions. 

To  prevent we, user can split subprograms a) and b) into two new separate nodeTrees called 4a and 5b, and then append only those two nodetrees. Now we don't have to append nodeTree1 and 2 if we don't need them. But building library of subprograms this way is bit clumsy IMO. 
Now how about marking (by user) some subprograms as 
-*_private *_(only for use in current node tree - not visible in 'shift+a'-> 'subprogram' for others), 
-and *_public *_(they are reusable subprograms that other node trees may use)? 

Could then maybe definitions of private subprograms be hidden in some 'Library' nodeTree, which would contain all public subprograms? But then if we want to append nodeTree XXX which is using some subprograms from 'Liblary', user would have to append whole 'liblary' thing, just to get required one or two subprograms definitions from it. 
Basically, in perfect world, user would just need to append desired node tree, and  private 'subprograms' would be included with it.  Just like with cycles materials - when you append material, nodeGroup definitions are imported with it. And blender is not so smart, to know he already have nodeGroup loaded, so with each material import, he creates nodeGroup.001, 002, etc.
I can't think of anything better.

**Comment by og76** [comment]
I don't really follow the private/public logic. Why would there be a subprogram that does not appear in subprograms? 

As for transferring the trees or subprograms library, there is always another way, that is having an **import/export in animation nodes**, not blender, **as json, xml, py** or another file, that will generate nodes/links **much like the templates**, full trees or parts of it. 

(This may avoid other confusions, like when you import an object and it comes with the material, it gets the node trees, but that is very complicated for objects affected by AN.)

**Comment by JoseConseco** [comment]
The whole thing is about simplifying work with multiple nodeTrees with subprograms. 
Private/Public thing is from C++ where there may be method in class, that you do not want to be seen outside class.
This is how I do things now in AN. When I have simple repetitive task I put it into separate node tree subprogram:
![image](https://cloud.githubusercontent.com/assets/13521338/13554222/a02330a8-e3a2-11e5-9e41-440a3c2c9135.png)
Lets call those as public 'snippets' that could be put into eg. shift+a menu - into new category 'snippet' or something. It can be accessed in any other node tree.
And below is example of tree that has subporgram marked in red, that is no really usefull outside this specific node tree:
![image](https://cloud.githubusercontent.com/assets/13521338/13554281/0444df0e-e3a4-11e5-872a-4ae9dc4f35ef.png)
This greenesh subprogram would be private - meaning not visible in 'snippets' menu.
The way I have setup everything now is bit hard to manage eg:
![image](https://cloud.githubusercontent.com/assets/13521338/13554304/cb441cc8-e3a4-11e5-915e-1da3bcb6e923.png)

NodeTrees marked in red  are not doing anything - they are just the 'snippets' templates. IMO, they could be removed from this list, and could be accessible from shift+a-> snippets menu (kind of similar to how templates in AN works now). Maybe those could be exported by xml, json as you mentioned.

The 'bricks' and 'suzaneCones' are the two actual node trees that are influencing the scene but the way it is setup now, everything is mixed together (meaning no separation of snippets/public templates from active nodeTrees which are generating output on scene).

**Comment by og76** [comment]
ok, I start getting the idea. I guess.

I see 3 situations here: 

1. used subprograms and nodes
2. not used, but present as nodes
3. presets/templates that are available but not existing as nodes

do we need to separate 1 and 2 ? how do we separate then 2 from 3 ?
That is how I see what you say. Is it?

-

I'm not a coder and C++ is far away, I do get this concept (probably), but I'd like to start this from the ux/ui pov rather than similarity with coding logic. 
_What do we gain with this vs how much it complicates An even further?_

Not sure I make sense above, but I said to just have a clear **1 file export/import in AN** for a template-like approach precisely to make a distinction between used and not used, but available preset/subprogram.

This would solve the issue of i/o and presets in **an expected, easy to understand way**.

Now, on this issue and the general library/import-export, should we rather look at it like normal blender users and not a coding thinking pov? I say this cause Loops are already a concept issue for a non coding/scripting user, a brand new ux in itself. And AN is a UI/workflow thing after all.

**Comment by JoseConseco** [comment]
Yes there are situations 1,2 and 3 as you mentioned. But I do not want to separate 1 and 2 from user pov. User should be just able to insert template -subprogram invoke node - created by himself previously and stored into json.  
If possible, nodes describing how template subprogram works - inner body of subprogram, should be hidden from user because he may not be interested how it works. 
How we separate between 1. and 2? - we don't and it works same as before. We are just interested in separating subprograms marked as 'Template'. From user pov:
- he right-click on invoke node for subprogram he wishes to convert to template. 
- menu pops (or pie) user chooses -Save Template- subprogram is saved into json on hard drive. Nodes defining template are hidden  from current nodeTree. But now new template is available from eg. Shift+a menu. I'm not sure is how we should hide nodes defining template from user. Is it possible to put them into hidden nodeTreeTemplateXYZ ?.
- when user want's to load old template he goes "shift+a"->templates-> Load templateABC. Node pops under cursor with TemplateABC invoke subprogram. Nodes defining body of TemaplateABC are hidden  (again not sure what is possible from blender API, but maybe creating hiden NodeTreeTemplateABC would work).

_What do we gain with this vs how much it complicates An even further?_
We gain: 
-less cluttered node trees (because Template subprograms are hidden (except invoke node))
-easy access to templates form shift+a menu in any blend file - cos templates are saved in json file on hard drive and not in blend file (this solves problem from first post where I had to append multiple nodeTrees, just to get all necessary subprograms, necessary to run another nodeTree). 
-possibility to share json file with others.
The down side i can see is, that if users deletes template - all blend files (or other templates) using this template may get broken (unless each blend file using template eg. ABC, would have hidden node tree 'NodeTreeABC ' with definition of subprogram nodes saved into blend file).  Or if user edits template, some node setups may get broken.  I suppose user should have options to : edit, create, delete, modify templates. And it is probably lots of work for Jacques :(.

**Comment by og76** [comment]
So basically you want to have a method to hide the subprogram definition. I guess the blender groups way is doing just that. By having another tree. The only difference is that script nodes do list all the trees and currently you need to import all trees manually (no automation here, yet).

For this, the template method or external file is not gonna work. It would be like translating a tree into a node or a script node or a plain py and:

- I think we are a long way from that. Not impossible, but far.
- I talk about the template/preset just as it works today, + as a method of exchange 
- templates/presets allow any node configuration, not just subprograms

-

Ok, could we separate the "i/o of trees and setups" from the actual "tree as in blender group" issue?

So, for the sake of the title "easy to reuse trees", I stand for the **template/preset as just an i/o** method alternative to normal blender append/import. Or for a library.

As for the "group as in blender", that is slightly different issue. my impression is that is a big effort with not so obvious results. Just an impression.

**Comment by JoseConseco** [comment]
> So basically you want to have a method to hide the subprogram definition
Well yes, but this is just bonus . I shouldn't have mentioned it at first position of benefits. The main thing I try to solve/ simplify is:
-easy way for storing and reusing subprograms, across multiple blend files (I renamed the title of this  thread because old was not precise). That is why I though json is good way to go - you store AN subprogram group/loop on hard drive. And AN can have access to it form any blend file, because they are not stored inside blend file itself. Eg. lets say I created 'randomize vector' subprogram group, and I know I will want to use it in future. The way I store it now is, I put it into separate NodeTree and store in blender startup file. But this way, I would end with many NodeTrees, just for purpose of storing subprograms.

> I talk about the template/preset just as it works today, + as a method of exchange
Current templates are ok, except 
When you append same template  multiple times to nodeTree, each time whole template is generated as separate copy. Meaning each subprogram is generated multiple times too. 
![image](https://cloud.githubusercontent.com/assets/13521338/13558495/18218a2e-e406-11e5-84d9-6d2d29091b17.png). 

> templates/presets allow any node configuration, not just subprograms
Well it is actually cool. 

> template/preset as just an i/o method alternative to normal blender append/import
Fine with me, I just wish, template didn't generated separate copy of whole stored node setup.

**Comment by JacquesLucke** [comment]
I thought about similiar things during the big refactor last summer. Not everything mentioned here is possible because of limitations in Blender and AN.

**Public and Private Subprograms**
This is definitly possible and quite easy to implement, it's just another property next to the subprogram name and description.
So the implementation is easy but the main question is if it makes AN better or worse. So I think you don't need it when you only have 5 subprograms or so, but I agree when you are making a library with multiple subprograms it can be quite useful! Unfortunally there are other problems when you make large node libraries.

**Problems with many node**
Currently everytime you change a link or add a node or so all node trees are recompiled. This is quite fast when you only have a few nodes but as soon as you have hundreds or nodes in your file this can noticeably slow down your workflow. The solution to this problem would be to only recompile the node tree that changed. Tthis would require some more refactoring. It is possible but I won't do that soon as it is only needed in extrem cases. 
If you know some python you can compress many nodes in one script node -> shorter compilation time.

When I say "compilation" I mean the process of converting the node trees into python code and convert that into CPython byte code.

**Converting nodes into standalone python code**
This is also something I thought about a lot but I decided that I want to keep this functionality out of AN for a couple of reasons.

- Nodes in AN serve multiple purposes. They handle data AND they store data. If nodes would only handle data it would be easier to create a standalone python script. But having actual nodes which can store data in the .blend file gives us much more possibilities. (eg the Object Instancer node has to store the object names)
- Bugs! Bugs always exist. No one can write code that doesn't have any bugs. Imagine you created a node tree and find a bug in one node. Currently you can just report it here, someone will fix it, you can download the new version and your node tree works again because the node simply uses the new code in the background. When you convert your node tree to python code beforehand and you find a bug there it is much harder to correct it.
- Many nodes use utility functions I wrote to simplify the AN development in the future and to make the code more readable. When you want to convert these nodes into a standalone script we have to copy these functions into the output script. This would result in even more "indirect coding" (not sure if this term exists, I mean to write code that will be copied into another context and executed then. This what I do in the getExecutionCode function all the time. It makes the execution much faster because it collapses abstractions, but it's a different kind of coding)

**Node Trees in .json or similiar files**
This is partially possible. As I have already explained we always need the actual nodes to be able to execute. "Virtual nodes" won't work because they can't store data in the .blend file.
In theory it is possible to make an operator that exports a nodetree but it is a bit harder than it seems at the first glimpse, because in AN you have to create the nodes in a certain order.
The easiest way to share your trees is simply to share the .blend file so that others can append the node tree. (also note: Linking is not supported because linked nodes can't store data because they can only be modified in the original file)
